tests/e2e/vmlistener_test.go

```diff
@@ -104,34 +116,85 @@ func (s *VMListenerTestParams) TestServiceRestart(c *C) {
+	c.Assert(isStatusChanged, Equals, true, Commentf("VM [%s] should be powered on state", s.vm1Name))
+
+	// Status should be detached
+	status := verification.VerifyDetachedStatus(s.volumeName, s.vm1, s.esx)
+	c.Assert(status, Equals, true, Commentf("Volume %s is still attached", s.volumeName))
 
-	// grab worldID/vmProcessID
-	processID := esxcli.GetVMProcessID(s.esxIP, s.dockerHostName)
-	log.Printf("VM's process ID is: %s", processID)
+	misc.LogTestEnd(c.TestName())
+}
 
-	// kill VM
-	isVMKilled := esxcli.KillVM(s.esxIP, processID)
-	c.Assert(isVMKilled, Equals, true, Commentf("Unable to kill VM %s ...", s.dockerHostName))
+// Test failover by killing a VM and then mount the volume to a different VM
+// 1. Setup: Create a volume
+// 2. Setup: Attach it to a container
+// 3. Kill the VM
+// 4. Verification: volume status should be detached
+// 5. Attach it to a different container
+// 6. Verification: volume status should attached
+func (s *VMListenerTestParams) TestFailover(c *C) {
+	misc.LogTestStart(c.TestName())
 
-	isStatusChanged := misc.WaitForExpectedState(govc.GetVMPowerState, s.dockerHostName, properties.PowerOffState)
-	c.Assert(isStatusChanged, Equals, true, Commentf("VM [%s] should be powered off state", s.dockerHostName))
+	// Kill the VM
+	isVMKilled := killVM(s.esx, s.vm1Name)
+	c.Assert(isVMKilled, Equals, true, Commentf("VM [%s] was not killed successfully", s.vm1Name))
 
-	// status should be detached
-	volStatus := verification.GetVMAttachedToVolUsingAdminCli(s.volumeName, s.esxIP)
+	// Status should be detached
+	volStatus := verification.GetVMAttachedToVolUsingAdminCli(s.volumeName, s.esx)
 	c.Assert(volStatus, Equals, properties.DetachedStatus, Commentf("Volume %s is still attached", s.volumeName))
 
-	// power on VM
-	govc.PowerOnVM(s.dockerHostName)
-	isStatusChanged = misc.WaitForExpectedState(govc.GetVMPowerState, s.dockerHostName, properties.PowerOnState)
-	c.Assert(isStatusChanged, Equals, true, Commentf("VM [%s] should be powered on state", s.dockerHostName))
+	// Attach to a container in a different VM
+	out, err := dockercli.AttachVolume(s.vm2, s.volumeName, s.containerName)
+	c.Assert(err, IsNil, Commentf(out))
 
-	// status should be detached
-	status := verification.VerifyDetachedStatus(s.volumeName, s.dockerHostIP, s.esxIP)
-	c.Assert(status, Equals, true, Commentf("Volume %s is still attached", s.volumeName))
+	// Status should be attached
+	status := verification.VerifyAttachedStatus(s.volumeName, s.vm2, s.esx)
+	c.Assert(status, Equals, true, Commentf("Volume %s is not attached", s.volumeName))
+
+	// Remove the container
+	out, err = dockercli.RemoveContainer(s.vm2, s.containerName)
+	c.Assert(err, IsNil, Commentf(out))
+
+	// Power on VM
+	govc.PowerOnVM(s.vm1Name)
+	isStatusChanged := misc.WaitForExpectedState(govc.GetVMPowerState, s.vm1Name, properties.PowerOnState)
+	c.Assert(isStatusChanged, Equals, true, Commentf("VM [%s] should be powered on state", s.vm1Name))
```

Comment:
Agree that there is an overlap. I had a discussion with @shuklanirdesh82 in such cases. We should have it for now since we do not want to miss any bug in any of the possible permutation.